error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ConstantNaming.java

```diff
@@ -0,0 +1,125 @@
+public final class ConstantNaming extends BugChecker implements VariableTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Matcher<VariableTree> IS_CONSTANT =
+      allOf(hasModifier(Modifier.STATIC), hasModifier(Modifier.FINAL));
+  private static final Matcher<VariableTree> IS_PRIVATE = hasModifier(Modifier.PRIVATE);
+  private static final Pattern SNAKE_CASE = Pattern.compile("([a-z])([A-Z])");
+  private static final ImmutableSet<String> DEFAULT_EXEMPTED_NAMES =
+      ImmutableSet.of("serialVersionUID");
+
+  /**
+   * Flag using which constant names that must not be flagged (in addition to those defined by
+   * {@link #DEFAULT_EXEMPTED_NAMES}) can be specified.
+   */
+  private static final String ADDITIONAL_EXEMPTED_NAMES_FLAG =
+      "CanonicalConstantNaming:ExemptedNames";
+
+  private final ImmutableSet<String> exemptedNames;
+
+  /** Instantiates a default {@link ConstantNaming} instance. */
+  public ConstantNaming() {
+    this(ErrorProneFlags.empty());
+  }
+
+  /**
+   * Instantiates a customized {@link ConstantNaming}.
+   *
+   * @param flags Any provided command line flags.
+   */
+  @Inject
+  ConstantNaming(ErrorProneFlags flags) {
+    exemptedNames =
+        Sets.union(DEFAULT_EXEMPTED_NAMES, Flags.getSet(flags, ADDITIONAL_EXEMPTED_NAMES_FLAG))
+            .immutableCopy();
+  }
+
+  @Override
+  public Description matchVariable(VariableTree tree, VisitorState state) {
+    String variableName = tree.getName().toString();
+    if (!IS_CONSTANT.matches(tree, state) || exemptedNames.contains(variableName)) {
+      return Description.NO_MATCH;
+    }
+
+    String replacement = toUpperSnakeCase(variableName);
+    if (replacement.equals(variableName)) {
+      return Description.NO_MATCH;
+    }
+
+    Description.Builder description = buildDescription(tree);
+    if (!IS_PRIVATE.matches(tree, state)) {
+      description.setMessage(
+          "%s; consider renaming to '%s', though note that this is not a private constant"
+              .formatted(message(), replacement));
+    } else if (isVariableNameInUse(replacement, state)) {
+      description.setMessage(
+          "%s; consider renaming to '%s', though note that a variable with this name is already declared"
+              .formatted(message(), replacement));
+    } else {
+      description.addFix(SuggestedFixes.renameVariable(tree, replacement, state));
+    }
+
+    return description.build();
+  }
+
+  private static String toUpperSnakeCase(String variableName) {
+    return SNAKE_CASE.matcher(variableName).replaceAll("$1_$2").toUpperCase(Locale.ROOT);
+  }
+
+  private static boolean isVariableNameInUse(String name, VisitorState state) {
+    return Boolean.TRUE.equals(
+        new TreeScanner<Boolean, @Nullable Void>() {
+          @Override
+          public Boolean visitVariable(VariableTree tree, @Nullable Void unused) {
+            return ASTHelpers.getSymbol(tree).getSimpleName().toString().equals(name)
```

Comment:
Can be simplified to
```suggestion
            return ASTHelpers.getSymbol(tree).contentEquals(name);
```

I noticed a similar discussion [here](https://github.com/PicnicSupermarket/error-prone-support/pull/783#discussion_r1502274717), so decided to create a Refaster rule: #1379.